### PR TITLE
composite-queries & variables

### PR DESCRIPTION
**Pull request by Amber1990Zhang**

### CODE REVIEW

**Review by whitewum — DISMISSED** [review]

**Comment by whitewum** [comment]
cypher composite is quite different. We can add cypher-composite later.

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Composite queries
+
+Composite queries put data from different queries together. They then use filters, group bys, or sorting before returning the combined return results. Composite Query retrieves multiple levels of related information on existing queries and presents data as a single return result.
```
Comment (randomJoe211) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md:3]: > Composite Query retrieves multiple levels of related information

A composite query

**Review by whitewum — APPROVED** [review]

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Composite queries
+
+Composite queries put data from different queries together. They then use filters, group bys, or sorting before returning the combined return results. A composite query retrieves multiple levels of related information on existing queries and presents data as a single return result.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md:3]: group bys? ===> GROUP BYs  or  group-bys ?

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+Nebula Graph supports two ways to compose queries (or sub-queries):
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md:5]: why use parentheses around "or sub-queries"?

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+- More than one queries can be batched together, separated by semicolon (;). The result of the last query is returned as the result of the batch.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md:7]: semicolons

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+- Queries can be piped together by using the pipe operator `|`. The result of the previous query can be used as the input of the next query.
+
+> **NOTE:** Composition queries are not `Transactional` queries.
+
+For example, a query composed of three sub-queries: A | B | C, In that A is a read operation, B is a computation operation, and C is a write operation. If any part fails in the execution, the whole result can be undefined. Currently, there is no roll back. What is written depends on the query executor.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md:12]: "A | B | C, In"==> period

Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/1.composite-queries.md:12]: roll back ==> rollback?

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# User-defined variables
+
+User-defined variables allows passing the result of one statement to another. User-defined variables are written as `$var_name`. The `var_name` consists of alphanumeric characters, any other characters are not recommended currently.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md:3]: "characters, any other characters" ==> "characters. Any other characters"? OR use conjunction.

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+User-defined variables can only be used in one execution. For example, you can use user-defined variables in composite queries separated by semicolon `'` or pipe `|`. Details about composite queries, see [Composite queries](1.composite-queries.md).
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md:5]: "semicolon `'`"?

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+> **NOTE:** A user-defined variable is valid only at the current session and execution.
+
+A user-defined variable in one statement can **NOT** be used in either other clients or other executions. The statement that defines the user-defined variable and the statement that uses it must be submitted together. When this session ends the user-defined variable is automatically expired.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md:9]: **CANNOT**

Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md:9]: When this session ends,

File: docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
++----------------------------------------------------+
+| v                                                  |
++----------------------------------------------------+
+| ("player100" :player{name: "Tim Duncan", age: 42}) |
++----------------------------------------------------+
+```
+
+The user-defined variable in the preceding query is `v`.
+
+> **NOTE:** This kind of variables are not carried over to subsequent queries.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/4.variable-and-composite-queries/2.user-defined-variables.md:41]: carried over ==> transferred?

File: mkdocs.yml [hunk 10]
```diff
@@ -75,6 +75,11 @@ nav:
         - Map: 3.ngql-guide/3.data-types/8.map.md
         - Type conversion: 3.ngql-guide/3.data-types/9.type-conversion.md
 
+      - Variables and composite variables:
```
Comment (oldLady344) [inline mkdocs.yml:78]: composite variables?

**Review by randomJoe211 — APPROVED** [review]
Only some minor problems. Please fix them later.

**Pull request merged**